### PR TITLE
Metric Testing + Implementation

### PR DESCRIPTION
**Pull request by aditya0by0**
- Pull Request for Issue number #22 

## Test the Macro-F1
- [x] Compare the result for iteratively updating (i.e., passing batches to the `update` function) and calculation via a single call
- [x] Compare the custom implementation to the torchmetrics `MultilabelF1Score`
- [x] Cover different datasets 
  - trivial cases: e.g. all predictions are 1 and half the labels are 1, all the labels are 1 and half the predictions are 1
  - case where some classes have no labels
  - realistic case with data from model output

## Add balanced accuracy metric
- [x] Add implementation of the balanced accuracy `ba = (tpr + tnr)/2 = (tp/(tp+fn) + tn(tn+fp))/2`
- [x] Add tests (similar to macro-f1)

### CODE REVIEW

File: tests/testCustomBalancedAccuracyMetric.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+import unittest
+import torch
+import os
+from chebai.callbacks.epoch_metrics import BalancedAccuracy
+import random
+
+
+class TestCustomMacroF1Metric(unittest.TestCase):
```
Comment (sfluegel05) [inline tests/testCustomBalancedAccuracyMetric.py:8]: This should be `TestBalancedAccuracyMetric`

File: tests/testCustomBalancedAccuracyMetric.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+        single_call_custom_metric = BalancedAccuracy(num_labels=num_labels)
+        single_call_custom_metric_score = single_call_custom_metric(preds, label).item()
+
+        self.assertEqual(iterative_custom_metric_score, single_call_custom_metric_score)
+
+    def test_metric_against_realistic_data(self):
+        """Test the custom metric against the standard on realistic data"""
+        directory_path = "CheBIOver100_test"
+        abs_path = os.path.join(os.getcwd(), directory_path)
```
Comment (sfluegel05) [inline tests/testCustomBalancedAccuracyMetric.py:35]: Could you add the data to the repository? (a few batches are probably sufficient)

File: tests/testCustomMacroF1Metric.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+
+class TestCustomMacroF1Metric(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    @unittest.expectedFailure
+    def test_all_predictions_are_1_half_labels_are_1(self):
```
Comment (sfluegel05) [inline tests/testCustomMacroF1Metric.py:16]: instead of excluding the test case, could you add assertions for the expected values?

File: chebai/callbacks/epoch_metrics.py [hunk 4]
```diff
@@ -47,3 +47,70 @@ def compute(self):
         # if (precision and recall are 0) or (precision is nan), set f1 to 0
         classwise_f1 = classwise_f1.nan_to_num()
         return torch.mean(classwise_f1)
+
+
+class BalancedAccuracy(torchmetrics.Metric):
```
Comment (sfluegel05) [inline chebai/callbacks/epoch_metrics.py:52]: Here, a comment with a short explanation would be nice

**Review by sfluegel05 — COMMENTED** [review]
Looks good so far, I added some comments to the code

File: tests/testCustomMacroF1Metric.py (hunk 3 above)
Comment (sfluegel05) [reply tests/testCustomMacroF1Metric.py:15]: @aditya0by0 Could you also do this for the other test cases that have the `expectedFailure` annotation

**Review by sfluegel05 — CHANGES_REQUESTED** [review]

**Review by sfluegel05 — APPROVED** [review]
